src/iop/dual_demosaic.c

```diff
@@ -0,0 +1,255 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2010-2020 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+   Dual demosaicing has been implemented by Ingo Weyrich <heckflosse67@gmx.de> for
+   rawtherapee under GNU General Public License Version 3
+   and has been modified to work for darktable by Hanno Schwalm (hanno@schwalm-bremen.de).
+   Also the code for fast_blur has been taken from rawtherapee capturesharpening,
+   implemented also by Ingo Weyrich.
+*/
+#ifdef __GNUC__
+  #pragma GCC push_options
+  #pragma GCC optimize ("-Ofast")
+#endif
+
+static INLINE float calcBlendFactor(float val, float threshold)
+{
+    // sigmoid function
+    // result is in ]0;1] range
+    // inflexion point is at (x, y) (threshold, 0.5)
+    return 1.0f / (1.0f + dt_fast_expf(16.0f - (16.0f / threshold) * val));
+}
+
+static void fast_blur(float *const restrict src, float *const restrict out, const int width, const int height, const float sigma)
+{
+  // For a blurring sigma of 2.0f a 13x13 kernel would be optimally required but the 9x9 is by far good enough here 
+  float kernel[9][9];
+  const double temp = -2.0f * sqrf(sigma);
+  float sum = 0.0f;
+  for(int i = -4; i <= 4; i++)
+  {
+    for(int j = -4; j <= 4; j++)
+    {
+      if( (sqrf(i) + sqrf(j)) <= sqrf(3.0 * 1.5) )
```

Comment:
@jenshannoschwalm 

Copy/paste bug: should be sqrf(3.0 * 2.0) because you call it with sigma 2.0

Reply:
just a note: i'd recommend writting that as `sqrtf(3.0f * 1.5f)` just so we're keeping with the idea of "we try to keep everything in float, unless there's clear need for double" :)

Reply:
I think it should be (3.0f * sigma)?

Not copy and paste but from testing for different sigmas :-)

Reply:
> I think it should be (3.0f * sigma)?

In RT it's 3.0 * 2.0 because 2.0 is the max sigma which is used for this kernel.

In your 9x9 kernel you could eliminate this condition because it always will be true

Reply:
ok & thanks for pointing out